### PR TITLE
ENH: Add tiff export function factory.

### PR DESCRIPTION
**Pull request by danielballan**
@sbillinge @timoHMM This is how I would implement TIFF export.

```python
In [2]: from bluesky.broker_callbacks import exporter_factory

In [3]: f = exporter_factory('img', 'test/{h.start.scan_id}_{N}.tiff')

In [4]: header = db['0b96536a']  # example data with 20 images in a field named `'img'`

In [6]: !mkdir test

In [7]: f(header)

In [8]: !ls test
1_0.tiff   1_11.tiff  1_14.tiff  1_17.tiff  1_2.tiff   1_5.tiff   1_8.tiff
1_1.tiff   1_12.tiff  1_15.tiff  1_18.tiff  1_3.tiff   1_6.tiff   1_9.tiff
1_10.tiff  1_13.tiff  1_16.tiff  1_19.tiff  1_4.tiff   1_7.tiff
```

As you can see, it is simple to customize the filename based on any piece of metadata. It's all contained in the Header. See docstring for more.

### CODE REVIEW

**Comment by danielballan** [comment]
cc @pavoljuhas

**Comment by pavoljuhas** [comment]
I think the example should say

```python
In [3]: f = exporter_factory('pe1_image_lightfield', 'test/{h.start.scan_id}_{N}.tiff')
```

Is this intended for end users, ie, people with a little Python/command-line knowledge, possibly stressing out at 3AM at the beamline?  If so I would recommend some bells and whistles from the [TiffSaver](https://github.com/NSLS-II-XPD/ipython_ophyd/blob/master/uiophyd/tiffsaver.py), i.e.,

* support for dry-run that would let users check output names before writing.
* exporter should have an attribute for filename-template that could be checked or changed without
  having to instantiate a new one.
* timestamping of the written tiff files by the scan time.
* ability to overwrite wrongly-named files identified by the same timestamp.
* dark image subtraction.  Possibly a boolean flag for carrying it out or not.

Is there a potential for the factory to support more output formats?  If it will be tiff-only, perhaps the factory might be renamed to `tiff_exporter_factory` or `make_tiff_exporter`.

**Comment by danielballan** [comment]
My example was applied to example data on my computer, generated from
databroker.examples.sample_data. Your example makes sense for XPD's pe1.

+1 on the rename and any bells, whistles people want. I'm just suggesting
this pattern.

**Comment by pavoljuhas** [comment]
Factory pattern is nifty and would be very useful if there were several exporter types, for example for several output formats.  I am however not sure what is its benefit for a single exporter type.  The factory call seems to be the only way of configuring the exporter and I don't see any way of accessing its configuration later.  It feels a bit clunky having to pass all of exporter parameters to the factory if user needs to change just one.  Am I perhaps missing something?

**Comment by danielballan** [comment]
Did you notice that the configuration is a template on `h`? If I call `f` repeatedly, using several different headers, each header's metadata fills in the template.

There are only two things to configure here -- which field (which in your case will probably never change) and which template, as in `'test/{h.start.scan_id}_{N}.tiff'`. I emphasize that `h` is a stand-in for any Header, not a specific Header. If you need to modify the template, you'd have to retype it anyway, right?

I don't expect non-sophisticated or overtired users will need to customize the template very often either, so all-purpose `f` could be defined in the IPython profile, ready to use. The simplicity of a function appeals to me.

File: bluesky/broker_callbacks.py [hunk 1]
```diff
@@ -87,3 +90,47 @@ def f(name, stop_doc):
...
+    """
+    # validate user input
+    if not '{N}' in template:
+        raise ValueError("template must include '{N}'")
+
+    def f(h):
+        imgs = get_images(h, field)
+        # Fill in h, defer filling in N.
+        _template = template.format(h=h, N='{N}')
+        for i, img in enumerate(imgs):
```
Comment (tacaswell) [inline bluesky/broker_callbacks.py]: This should generate and check if any of the files exist.  If _any_ exist it should raise.

**Comment by danielballan** [comment]
New commits add a `dryrun` argument to the exporter function. Also, the function now always returns a list of filepaths to the files it wrote (or, if dryrun *would* write).

I like the idea of the timestamps -- can you open up a PR to incorporate that since you've done it before?

The rest, particularly exporting *processed* data, is a more complex task, out of scope for this light-weight utility. The functionality is needed, but I'd like to build it separately.

**Comment by danielballan** [comment]
P.S. One useful thing I want to highlight: it is possible to hook this up to bluesky subscriptions such that TIFFs are automatically exported at the end of a scan. Users don't have to intervene or know anything about how it works.

**Comment by danielballan** [comment]
Useful notebook. I already sent this over email, but I'm posting the link on github so we can find it later. http://nbviewer.ipython.org/gist/anonymous/a3ee62067986e73b2d2b

**Comment by tacaswell** [comment]
That notebook should go into our examples or something.

**Comment by danielballan** [comment]
See #13

**Comment by pavoljuhas** [comment]
Took a while to get to this.  See #263.

**Pull request merged**